docs/tp/01.md

```diff
@@ -2,4 +2,486 @@
+
+```rust
+fn build_user(email: String, username: String) -> User {
+    User {
+        active: true,
+        username: username,
+        email: email,
+        sign_in_count: 1,
+    }
+}
+```
+
+#### Tuple structs
+Rust prend également en charge les structures qui ressemblent aux tuples, appelées **tuple structs**. Les structures de tuple ont la signification supplémentaire fournie par le nom de la structure mais n'ont pas de noms associés à leurs champs ; au lieu de cela, ils ont juste les types des champs. Les structures de tuple sont utiles lorsque vous souhaitez donner un nom à l'ensemble du tuple et lui donner un type différent des autres tuples, et lorsque nommer chaque champ comme dans une structure régulière serait verbeux ou redondant.
+
+```rust
+struct Color(i32, i32, i32);
+struct Point(i32, i32, i32);
+ 
+fn main() {
+    let black = Color(0, 0, 0);
+    let origin = Point(0, 0, 0);
+}
+```
+:::tip 
+Pour comprendre mieux, lire le **chapitre 5** qui se trouve dans la **documentation** au début du TP!
+:::
+
+### Enums
+Les énumérations, également appelées `enum`, vous permettent de définir un type en énumérant ses variantes possibles.
+Comment définir une énumération :
+
+```rust
+enum IpAddrKind {
+    V4,
+    V6,
+}
+```
+
+### String
+
+Rust n'a qu'un seul type de chaîne dans le langage principal, qui est la tranche de chaîne `str` qui est généralement vue sous sa forme empruntée `&str`.
+
+Le type `String`, qui est fourni par la bibliothèque standard Rust plutôt que codé dans le langage principal, est un type de chaîne codé en UTF-8 évolutif, mutable et détenu.
+
+#### Créer une nouvelle String
+```rust
+    let mut s = String::new();
+```
+Cette ligne crée une nouvelle chaîne vide appelée `s`, dans laquelle nous pouvons ensuite charger des données.
+
+Nous pouvons utiliser la fonction `String::from` ou la fonction `to_string` pour créer une chaîne à partir d'une chaîne littérale :
+
+```rust
+    let s = String::from("initial contents");
+```
+```rust
+let data = "initial contents";
+ 
+    let s = data.to_string();
+ 
+    // the method also works on a literal directly:
+    let s = "initial contents".to_string();
+```
+
+#### Ajouter des données à une string
+Nous pouvons développer une chaîne en utilisant la méthode `push_str` pour ajouter une tranche de chaîne.
+
+```rust
+let mut s = String::from("foo");
+s.push_str("bar");
+```
+
+**La méthode push** prend **un seul caractère** comme paramètre et l'ajoute à la chaîne.
+
+```rust
+    let mut s = String::from("lo");
+    s.push('l');
+```
+
+#### Méthodes d'itération sur les Strings
+La meilleure façon d’opérer sur des morceaux de chaînes est d’indiquer explicitement si vous voulez des caractères ou des octets. Pour les valeurs scalaires Unicode individuelles, utilisez la méthode `chars`.
+```rust
+for c in "Зд".chars() {
+    println!("{c}");
+}
+```
+
+#### Exécuter le programme
+On doit acceder a la directeur ou le fichier main.rs se trouve et executee la comande:
```

Comment:
This can be run anywhere from the crate's folder.

In order to run the program we may be anywhere in the crate's folder and execute the command.